translationUtils.ts

```diff
@@ -0,0 +1,284 @@
+import fetch from 'node-fetch';
+import imgur from 'imgur';
+import { Image, createCanvas } from 'canvas';
+
+imgur.setClientId('[REDACTED]');
+// This might not be necessary according to the docs, but can't hurt.
+imgur.setCredentials('[REDACTED]', '[REDACTED]', '[REDACTED]');
+
+type ImageBase64 = string; // Eg 'data:image/png;base64,iVBORw0KGgo...'
+
+interface TranslationResult {
+	originalLanguage: string;
+	translatedText: string;
+	minX: number;
+	minY: number;
+	maxX: number;
+	maxY: number;
+}
+
+enum StatusCode {
+	Ok = 200
+}
+
+// Returns null if the image could not be uploaded.
+export async function uploadImage(image: ImageBase64): Promise<URL | null> {
+	const formattedBase64 = image.replace(/^data:.*;base64,/g, '');
+	return new Promise(resolve =>
+		imgur
+			.uploadBase64(formattedBase64)
+			.then(function (json) {
+				const url = tryParseUrl(json.data.link);
+				resolve(url);
+			})
+			.catch(function () {
+				// Imgur can be down due to maintenance, high load, or other reasons.
+				// It is not an exceptional event, so simply return null.
+				resolve(null);
+			})
+	);
+}
+
+// Returns null if url is malformed.
+export function tryParseUrl(url: string): URL | null {
+	try {
+		return new URL(url);
+	} catch {
+		return null;
+	}
+}
+
+// Returns null if the image cannot be translated.
+export async function translate(image: ImageBase64): Promise<ImageBase64 | null> {
```

Comment:
Interesting historical note: this used to be `function translate(url: URL);` but I changed it to this.

The reason was- I needed a way to _not_ translate small images. There was a few things I could have done to achieve this.

- Option 1: change the function to `function translate(url: URL, requirements: { heightMin; widthMin });` that would only translate the image if it met the requirements.
- Option 2: change the function to `function translate(url: URL, verify: (image) => boolean);` that would invoke the callback `verify` to check if it should continue on and translate after fetching the image data.
- Option 3: Nothing! Keep it as `function translate(url: URL)` and simply hide the min height and width requirements in the implementation.
- Option 4: what I went with - decompose the function into modular blocks _even more_. `function translate(url: URL)` becomes `function translate(image: ImageBase64)` and it's up to the consumers to pass in what they want, vs `translate` fetching it for them.

Option 3 seems the worst to me, we're essentially hiding business logic of _one consumer_ in the implementation. This is mixing business logic with functional logic and may blow up in our face someday if we do want to translate smaller images (because future devs may not realize this hidden time bomb).

Option 1 and 2 are totally fine, but feel like they're doing too much. We're not simply "translating" anymore, we're translating _and_ checking. Option 1 or 2 are very similar, however my gut feeling likes Option 2 because it's more flexible and it will grow more organically. You've probably seen some function that takes a `config` object that has dozens of options before. This is because it's trying to be _everything_ to _everyone_. The inversion of responsibilities has set it up to be an endlessly growing blackhole of requirements. Instead, option 2 simply gives the consumers the object and lets them pick how they want to validate. It will scale much more efficiently. It can even export some common validators for usual use cases.

>Generality: Do Not Restrict What Is Inherent. Generality is the ability to use a function for many ends. It expresses the professional humility of the designer, his conviction that users will be inventitive beyond his imagination.
>Paraphrased from Fred Brook's, The Design of Design

Option 4 seemed to be the best idea to me. It kept the logic simple, the boundaries consistent, and the reusability high. When you encounter some function that doesn't support your functionality _dont_ just add a boolean flag- decompose the function more!